spec/models/work_packages/pdf_export/work_package_to_pdf_spec.rb

```diff
@@ -0,0 +1,120 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require 'spec_helper'
+require 'pdf/inspector'
+
+RSpec.describe WorkPackage::PDFExport::WorkPackageToPdf do
+  include Redmine::I18n
+  let(:type) { create(:type_bug) }
+  let(:project) { create(:project, name: 'Foo Bla. Report No. 4/2021 with/for Case 42', types: [type]) }
+  let(:user) do
+    create(:user,
+           member_in_project: project,
+           member_with_permissions: %w[view_work_packages export_work_packages])
+  end
+  let(:export_time) { DateTime.new(2023, 6, 30, 23, 59) }
+  let(:export_time_formatted) { format_time(export_time, true) }
+  let(:image_path) { Rails.root.join("spec/fixtures/files/image.png") }
+  let(:image_attachment) { Attachment.new author: user, file: File.open(image_path) }
+  let(:attachments) { [image_attachment] }
+  let(:work_package) do
+    description = <<~DESCRIPTION
+      **Lorem** _ipsum_ ~~dolor~~ `sit` [amet](https://example.com/), consetetur sadipscing elitr.
+      <mention data-text="@OpenProject Admin">@OpenProject Admin</mention>
+      ![](/api/v3/attachments/#{image_attachment.id}/content)
+      <p class="op-uc-p">
+        <figure class="op-uc-figure">
+          <div class="op-uc-figure--content">
+            <img class="op-uc-image" src="/api/v3/attachments/#{image_attachment.id}/content" alt='"foobar"'>
+          </div>
+          <figcaption>Image Caption</figcaption>
+         </figure>
+      </p>
+      <p><unknown-tag>Foo</unknown-tag></p>
+    DESCRIPTION
+    create(:work_package,
+           project:,
+           type:,
+           subject: 'Work package 1',
+           story_points: 1,
+           description:).tap do |wp|
+      allow(wp)
+        .to receive(:attachments)
+              .and_return attachments
+    end
+  end
+  let(:options) { {} }
+  let(:export) do
+    login_as(user)
+    described_class.new(work_package, options)
+  end
+  let(:export_pdf) do
+    Timecop.freeze(export_time) do
+      export.export!
+    end
+  end
+
+  def column_title(column_name)
+    label_title(column_name).upcase
+  end
+
+  def label_title(column_name)
+    WorkPackage.human_attribute_name(column_name)
+  end
+
+  subject(:pdf) do
+    content = export_pdf.content
+    # File.binwrite('WorkPackageToPdf-test-preview.pdf', content)
+    { strings: PDF::Inspector::Text.analyze(content).strings,
+      images: PDF::Inspector::XObject.analyze(content).page_xobjects.flat_map do |o|
+        o.values.select { |v| v.hash[:Subtype] == :Image }
+      end }
```

Comment:
❓ In the other test we are writing the PDF to disk, here we are checking the content directly. Can't we use the same procedure in the other file as well?

Reply:
The (other) query based export does have a feature to batch exporting thousands of work packages (with images) to different pdfs and concats them in the end with an external tool. Granted, the batching is currently not triggered in the spec, but this change wouldn't have much benefits.

Reply:
👍